### ISSUE TITLE
Provide NFC Library

### ISSUE THREAD

**Issue by erwango** [zephyrproject-rtos/zephyr#5703]
NFC is a important part of IoT ecosystem and now part of some basic use cases that should be supported by Zephyr.
It is also available on some boards and ICs supported by Zephyr.
A NFC library ported onto Zephyr would allow to extend Zephyr offer and provide more complete use cases.

First priority would be NFC Tag functionality would be sufficient. NFC reader capability would be a lesser prio considering targeted use case by Zephyr devices.

First step of the study would be to establish requested services level (from basic memory write to full NDEF support).

**Comment by rodrigopex** [comment]
@nashif any progress on that? Can I start to work to create the API/Driver?

**Comment by nullr0ute** [comment]
@rodrigopex I would suggest just get going on it if you have the cycles

**Comment by rodrigopex** [comment]
What do you mean with cycles?

**Comment by nullr0ute** [comment]
If you have the time to work on it (the cycles) I'd probably just do so, maybe check on IRC

**Comment by grochu** [comment]
> Guys, any progress on that? We will need NFC for a project. I am available to help on that if you can support me (information and help) on creating it. The first I will focus initially on the driver for nRF52832 device.

We are working on adding an NFC driver supporting nRF52832 and nRF52840 chips to the NRFX driver package (and consequently to Zephyr). The timeline is around end of September - beginning of October.

**Comment by rodrigopex** [comment]
Great! If you need help to test it when ready, let me know.

Best regards,
Rodrigo Peixoto

**Comment by erwango** [comment]
> We are working on adding an NFC driver supporting nRF52832 and nRF52840 chips to the NRFX driver package (and consequently to Zephyr).

Might be worth sharing the API once you have something ready. The sooner the API is discussed the better IMHO.

**Comment by grochu** [comment]
> Might be worth sharing the API once you have something ready. The sooner the API is discussed the better IMHO.

Agree, we can start discussing the API once it gets stable, but let me clarify 2 things already:

1. What we are going to provide in the first step is a HW driver for the nRF52 chips, so roughly it will cover what is supported by the NFC peripheral in these MCUs. We are not attempting to propose a general NFC driver API for Zephyr in this step.
2. To reveal some details on the driver API, I can say it is rather low-layer and corresponds to chapter 6 of the "NFC Forum Digital Protocol Technical Specification 2.0" (protocol layer equivalent to ISO 14443-3: _Initialization and anti-collision_).

**Comment by aaronhshi** [comment]
@grochu is proposing/developing a general NFC driver API for Zephyr in your guy's plan for the future? This issue is something I'm very interested in and was hoping to develop and be apart of in providing a general NFC driver API for Zephyr!

**Comment by grochu** [comment]
@aaronhshi yes, we are planning to propose a general NFC API for Zephyr as well. We will come up with an API-only pull request to start the discussion.

**Comment by thomasthep-depone** [comment]
Any updates regarding the status of NFC?

There's a pull-request https://github.com/zephyrproject-rtos/zephyr/pull/10117 which has since last been touched on April the 30th. It's been marked stale since.

**Comment by kshitizdange** [comment]
Any update on this? 
I do see that the `nfc_hello` sample that existed in `2.3.0` release, has been removed from `2.4.0` release.

**Comment by selman-nus** [comment]
> > Guys, any progress on that? We will need NFC for a project. I am available to help on that if you can support me (information and help) on creating it. The first I will focus initially on the driver for nRF52832 device.
> 
> We are working on adding an NFC driver supporting nRF52832 and nRF52840 chips to the NRFX driver package (and consequently to Zephyr). The timeline is around end of September - beginning of October.

@grochu Is there any snippet yet?

**Comment by grochu** [comment]
There is an NFCT driver in the hal/nordic repo: _modules/hal/nordic/nrfx/drivers/src/nrfx_nfct.c_ which supports nRF52 and nRF53 families (devices which have NFCT). At the moment there are no plans of adding a generic Zephyr NFC driver or upper-layer libraries, but we are open to support such activity.

**Comment by adamfowleruk** [comment]
Happy to see the Nordic nRF52 and 53 supported. Wanted to flag that this may be needed by the Herald Project for Digital Contact Tracing wearable checking-in, or out of band pairing for bluetooth smartphone with wearable (if possible... not sure if any SOCs have readers...), or by a number of covid credentials projects. Our project uses nRF52 and 53 chips as our reference hardware platform so with the comment from @grochu I think we are fine for now.

**Comment by ck-telecom** [comment]
any update for generic NFC library? could we steal from MBed?

**Comment by matt-wood-ct** [comment]
Is anyone working on this, I could really use it

**Comment by marcel-ithinx** [comment]
I'm working on a driver for the NXP CLRC661 and CLRC663 and ISO15693 and NXP slix protocol layers. All of that is still in work or internal review. I'll try to open a PR around mit october.

**Comment by pdgendt** [comment]
Hi @marcel-ithinx, I've also worked on an NFC proposal (which is far from complete), but maybe can help out for more ideas:
https://github.com/zephyrproject-rtos/zephyr/compare/main...pdgendt:zephyr:nfc-subsys

I've added the NXP PN5180 driver and ISO14443A subsystem, and tested both reader mode and HCE (Host Card Emulation).

You (or anyone) can ping me if desired :)

**Comment by MarcelKr** [comment]
(switching to proper account, oops)

@pdgendt Uh, that looks great! I didn't implement the NFC driver API abstraction (nfc.h) yet (which you already did). I'm on vacation the next two weeks but after that will have a closer look and probably will try to adapt my stuff to use your nfc.h. Thanks for your response :)

**Comment by HesselM** [comment]
Hi all! 

Last year I've been working with NFC on both nRF and NXP platforms. Part of our zephyr port to the qn9090 I actually made a NFC (tag) subsystem, supporting the nRF-NFCT drivers and the nth3x-nfc tag (which is internally located within the qn9090 die). 

I never bothered making it public, but as part of the qn9090 zephyr 3.4 port, i moved it to public space.
You can find my approach at : https://github.com/Sendrato/zephyr/tree/v3.4-branch-nfc.

The approach is 90% based on the approach nordic took. 
The nordic-drivers are probably outdated as they are from June 2021 or so?

**Comment by HesselM** [comment]
Hi all, 

An update: I've patches bugs, cleaned the branch/commit and upgraded all nRF drivers. 
The proposed NFC subsystem can be checked at https://github.com/Sendrato/zephyr/tree/v3.4-branch-nfc

I've included a sample which should run on every nordic device with NFC and devices using the nt3h2 driver.
A board-configuration-overlay for the dk6_qn9090 and mg100 are included.

Hessel.

**Comment by matt-wood-ct** [comment]
This is still interesting, any movement on this?

**Comment by gregistech** [comment]
> This is still interesting, any movement on this?

**Comment by matt-wood-ct** [comment]
I do still feel this is essentially a necessary addition, why can't we just adopt Nordics NFC implementation they wrote for supporting the NFC-A peripherals on their chips, I understand the current files have a Nordic license on, are they actually opposed to allowing its use on non Nordic platforms? Is there a wider issue at play?

**Comment by razvanphp** [comment]
@HesselM do you think we could work together to split your PR into a generic NFC subsystem? We would like to add PN7160 support via NCI 2.0 as well. Seem from the supported products list that the many NXP chips should work with a common driver: https://www.nxp.com/applications/technologies/security/industrial-security/nfc-reader-library-software-support-for-nfc-frontend-solutions:NFC-READER-LIBRARY

given "NXP is a founding and platinum member of the Zephyr Project", I think they should support this contribution.

**Comment by ShahinHaque97** [comment]
Has there been any movement on this?

**Comment by kshypachov** [comment]
Hi, are there any updates regarding the NFC subsystem?

I have a PN532/C1 chip for which I’m planning to implement a driver. I could contribute it to the Zephyr tree if there is already some understanding of how the NFC subsystem and its API should be structured.

**Comment by juanjin-dev** [comment]
> Hi, are there any updates regarding the NFC subsystem?
> 
> I have a PN532/C1 chip for which I’m planning to implement a driver. I could contribute it to the Zephyr tree if there is already some understanding of how the NFC subsystem and its API should be structured.

I also have a PoC built on @pdgendt's NFC API draft that has a PN53x driver in it, in case any of it is useful to you - https://github.com/zephyrproject-rtos/zephyr/compare/main...juanjin-dev:zephyr:subsys-nfc-poc. I haven't finished hardware bring-up yet though, so please don't treat it as a finished driver.

@pdgendt I actually have a question for you about that one.

I've kept your commit with you as the author and your Signed-off-by on it, and your copyright header is still on the files that came out of your work. The thing is I've changed it quite a lot since then. nfc.h is +431/-99 against your version now, mostly listener side operations, support for controllers that resolve and activate targets in firmware, and a property interface. At that point it isn't really your commit any more, and it didn't feel right to decide that by myself. Would you rather I keep it as one commit with your name on it, or split it so your original stays as it was and my changes go into a separate commit of mine? Happy to do it either way, just let me know which you prefer.

I should also say I'm aware of #91254 and that you're reviewing there. I'm not trying to go around it. Once I have hardware results I'll raise the overlap in that thread rather than quietly opening a competing proposal. I just wanted to sort out the authorship question with you first.
